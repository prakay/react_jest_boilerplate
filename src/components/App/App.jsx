import React, { Component, Fragment } from 'react';
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import nanoid from 'nanoid';
import { remove, findIndex, propEq } from 'ramda';
import 'react-tabs/style/react-tabs.css';

import './App.css';

class App extends Component {
  constructor(props) {
    super(props);
    const key = nanoid();
    this.state = {
      tabs: [
        { title: 'Tab', content: 'Content tab', key }
      ],
    };
  }

  setStateWithLog(state) {
    console.log('state', state);
    this.setState(state);
  }


  deleteTab(key) {
    const { tabs } =  this.state;
    const tabIndex = findIndex(propEq('key', key), tabs);
    if (tabIndex > -1) {
      this.setStateWithLog({ tabs: remove(tabIndex, 1, tabs) });
    }
  }

  addTab() {
    const { tabs } =  this.state;
    const key = nanoid();
    this.setStateWithLog({ tabs: [...tabs, { title: 'Tab', content: 'Content tab', key } ] })
}

  render() {
    const { tabs } = this.state;

    if (!tabs) return null;

    return (
      <Fragment>
        <Tabs>
          <TabList>
            {tabs.map((tab) => (
              <Tab data-test-name="mario-tab" key={tab.key}>
                {tab.title}
                <button type="button" onClick={() => this.deleteTab(tab.key)} data-test-name="mario-tab-delete">remove tab</button>
              </Tab>
            ))}
          </TabList>
          {tabs.map((tab) => (
            <TabPanel key={tab.key}>{tab.content}</TabPanel>
          ))}
        </Tabs>
        <button type="button" onClick={() => this.addTab()} data-test-name="mario-tab-add">Add tab</button>
      </Fragment>
    );
  }
}

export default App;
