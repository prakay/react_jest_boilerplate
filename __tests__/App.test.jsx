import React from 'react';
import { mount } from 'enzyme';

import App from '../src/components/App';

describe('test <App />', () => {
  it('render default tabs', () => {
    const wrapper = mount(<App />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('add new tab', () => {
    const wrapper = mount(<App />);
    const addtab = wrapper.find('[data-test-name="mario-tab-add"]');
    expect(wrapper.render()).toMatchSnapshot();
    addtab.simulate('click');
    expect(wrapper.render()).toMatchSnapshot();
  });
});
