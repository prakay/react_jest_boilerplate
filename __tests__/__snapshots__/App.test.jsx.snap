// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <App /> add new tab 1`] = `
Array [
  <div
    class="react-tabs"
    data-tabs="true"
  >
    <ul
      class="react-tabs__tab-list"
      role="tablist"
    >
      <li
        aria-controls="react-tabs-3"
        aria-disabled="false"
        aria-selected="true"
        class="react-tabs__tab react-tabs__tab--selected"
        data-test-name="mario-tab"
        id="react-tabs-2"
        role="tab"
        tabindex="0"
      >
        Tab
        <button
          data-test-name="mario-tab-delete"
          type="button"
        >
          remove tab
        </button>
      </li>
    </ul>
    <div
      aria-labelledby="react-tabs-2"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-3"
      role="tabpanel"
    >
      Content tab
    </div>
  </div>,
  <button
    data-test-name="mario-tab-add"
    type="button"
  >
    Add tab
  </button>,
]
`;

exports[`test <App /> add new tab 2`] = `
Array [
  <div
    class="react-tabs"
    data-tabs="true"
  >
    <ul
      class="react-tabs__tab-list"
      role="tablist"
    >
      <li
        aria-controls="react-tabs-3"
        aria-disabled="false"
        aria-selected="true"
        class="react-tabs__tab react-tabs__tab--selected"
        data-test-name="mario-tab"
        id="react-tabs-2"
        role="tab"
        tabindex="0"
      >
        Tab
        <button
          data-test-name="mario-tab-delete"
          type="button"
        >
          remove tab
        </button>
      </li>
      <li
        aria-controls="react-tabs-5"
        aria-disabled="false"
        aria-selected="false"
        class="react-tabs__tab"
        data-test-name="mario-tab"
        id="react-tabs-4"
        role="tab"
      >
        Tab
        <button
          data-test-name="mario-tab-delete"
          type="button"
        >
          remove tab
        </button>
      </li>
    </ul>
    <div
      aria-labelledby="react-tabs-2"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-3"
      role="tabpanel"
    >
      Content tab
    </div>
    <div
      aria-labelledby="react-tabs-4"
      class="react-tabs__tab-panel"
      id="react-tabs-5"
      role="tabpanel"
    />
  </div>,
  <button
    data-test-name="mario-tab-add"
    type="button"
  >
    Add tab
  </button>,
]
`;

exports[`test <App /> render default tabs 1`] = `
Array [
  <div
    class="react-tabs"
    data-tabs="true"
  >
    <ul
      class="react-tabs__tab-list"
      role="tablist"
    >
      <li
        aria-controls="react-tabs-1"
        aria-disabled="false"
        aria-selected="true"
        class="react-tabs__tab react-tabs__tab--selected"
        data-test-name="mario-tab"
        id="react-tabs-0"
        role="tab"
        tabindex="0"
      >
        Tab
        <button
          data-test-name="mario-tab-delete"
          type="button"
        >
          remove tab
        </button>
      </li>
    </ul>
    <div
      aria-labelledby="react-tabs-0"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-1"
      role="tabpanel"
    >
      Content tab
    </div>
  </div>,
  <button
    data-test-name="mario-tab-add"
    type="button"
  >
    Add tab
  </button>,
]
`;
